flang/test/Lower/OpenMP/omp-parallel-firstprivate-clause.f90

```diff
@@ -3,39 +3,36 @@
-!FIRDialect:    omp.terminator
-!FIRDialect:  }
+!FIRDialect: func @_QPfirstprivate_clause(%[[ARG1:.*]]: !fir.ref<i32>, %[[ARG2:.*]]: !fir.ref<!fir.array<10xi32>>, %[[ARG3:.*]]: !fir.ref<!fir.box<!fir.heap<i32>>>) {
+!FIRDialect-DAG:  omp.parallel
+!FIRDialect-DAG: %[[ARG1_PVT:.*]] = fir.alloca i32 {{{.*}}uniq_name = "{{.*}}Earg1"}
+!FIRDialect-DAG: %[[ARG1_PVT_LOAD:.*]] = fir.load %[[ARG1]] : !fir.ref<i32>
+!FIRDialect-DAG: fir.store %[[ARG1_PVT_LOAD]] to %[[ARG1_PVT]] : !fir.ref<i32>
+!FIRDialect-DAG: %[[ARG2_PRIVATE:.*]] = fir.alloca !fir.array<10xi32> {{{.*}}uniq_name = "{{.*}}Earg2"}
+!FIRDialect-DAG: %[[SHAPE:.*]] = fir.shape %c10 : (index) -> !fir.shape<1>
+!FIRDialect-DAG: %[[ARG2_INIT_VAL:.*]] = fir.array_load %[[ARG2_PRIVATE:.*]](%[[SHAPE]]) : (!fir.ref<!fir.array<10xi32>>, !fir.shape<1>) -> !fir.array<10xi32>
+!FIRDialect-DAG: %[[FIRST_PRIVATE_UPDATES:.*]] = fir.do_loop
+!FIRDialect-DAG: {{.*}}fir.array_fetch{{.*}}
+!FIRDialect-DAG: {{.*}}fir.array_update{{.*}}
+!FIRDialect-DAG: fir.result{{.*}}
+!FIRDialect-DAG: }
+!FIRDialect-DAG: fir.array_merge_store %[[ARG2_INIT_VAL]], %[[FIRST_PRIVATE_UPDATES]] to %[[ARG2_PRIVATE]] : !fir.array<10xi32>, !fir.array<10xi32>, !fir.ref<!fir.array<10xi32>>
+!FIRDialect-DAG: %[[ARG3_BOX:.*]] = fir.load %[[ARG3]] : !fir.ref<!fir.box<!fir.heap<i32>>>
+!FIRDialect-DAG: %[[ARG3_ADDR:.*]] = fir.box_addr %[[ARG3_BOX]] : (!fir.box<!fir.heap<i32>>) -> !fir.heap<i32>
+!FIRDialect-DAG: %[[ARG3_PVT:.*]] = fir.alloca i32 {{{.*}}uniq_name = "{{.*}}Earg3"}
+!FIRDialect-DAG: %[[ARG3_PVT_LOAD:.*]] = fir.load %[[ARG3_ADDR]] : !fir.heap<i32>
+!FIRDialect-DAG: fir.store %[[ARG3_PVT_LOAD]] to %[[ARG3_PVT]] : !fir.ref<i32>
+!FIRDialect-DAG: omp.terminator
+!FIRDialect-DAG: }
 
-!LLVMDialect: llvm.func @_QPfirstprivate_clause(%[[ARG1:.*]]: !llvm.ptr<i32>, %[[ARG2:.*]]: !llvm.ptr<array<10 x i32>>) {
-!LLVMIRDialect-DAG:  %[[ALPHA:.*]] = llvm.alloca %{{.*}} x i32 {{{.*}}, uniq_name = "{{.*}}Ealpha"} : (i64) -> !llvm.ptr<i32>
-!LLVMIRDialect-DAG:  %[[BETA:.*]] = llvm.alloca %{{.*}} x i32 {{{.*}}, uniq_name = "{{.*}}Ebeta"} : (i64) -> !llvm.ptr<i32>
-!LLVMIRDialect-DAG:  %[[GAMA:.*]] = llvm.alloca %{{.*}} x i32 {{{.*}}, uniq_name = "{{.*}}Egama"} : (i64) -> !llvm.ptr<i32>
-!LLVMIRDialect-DAG: %[[ALPHA_ARRAY:.*]] = llvm.alloca %{{.*}} x !llvm.array<10 x i32> {{{.*}}, uniq_name = "{{.*}}Ealpha_array"} : (i64) -> !llvm.ptr<array<10 x i32>>
-!LLVMIRDialect:  omp.parallel firstprivate(%[[ALPHA]] : !llvm.ptr<i32>, %[[BETA]] : !llvm.ptr<i32>, %[[GAMA]] : !llvm.ptr<i32>,
-!%[[ALPHA_ARRAY]] : !llvm.ptr<array<10 x i32>>, %[[ARG1]] : !llvm.ptr<i32>, %[[ARG2]] : !llvm.ptr<array<10 x i32>>) {
-!LLVMIRDialect:    omp.terminator
-!LLVMIRDialect:  }
-
-subroutine firstprivate_clause(arg1, arg2)
+subroutine firstprivate_clause(arg1, arg2, arg3)
 
         integer :: arg1, arg2(10)
-        integer :: alpha, beta, gama
-        integer :: alpha_array(10)
+        integer, allocatable :: arg3
```

Comment:
Only integer scalar, integer array, and allocatable array is tested here. What about other data types such as real, complex, quad, character, pointer, polymorphic without allocatable attribute, etc? Is it because these data types are still not supported in FIR? Please forgive me that I would like to know the status of the data types. I think implementation for different data types is relevant to `private`, `firstprivate`, `lastprivate` and `threadprivate`.

Reply:
Thanks @PeixinQiao. Yes, all data types are important. I believe this patch will work for real and complex also. I will add tests for these.
This patch is to get started. We will cover everything when we finish privatisation.

Reply:
OK to me.
Nit: I am not familiar the rules of adding test cases for the code implementation under `flang` directory. Should we add the different kinds for those scalar variables? They check if we transform the data type size correctly.